File: ml-agents-dots-envs~/mlagents_dots_envs/shared_memory/base_shared_memory.py
from abc import ABC
import os
import tempfile
import mmap
import numpy as np
import struct
import uuid
from typing import Tuple


class BaseSharedMemory(ABC):
    DIRECTORY = "ml-agents"

    def __init__(self, file_name: str, create_file: bool = False, size: int = 0):
        """
        Creates a bare bone shared memory wrapper that connects or creates a shared
        memory file.
        :string file_name: The name of the file used for shared memory
        :bool create_file: If true, the file will be created (any existing file with
        that name will be deleted).
        :int size: When creating the file, specifies its length in bytes
        """
        directory = os.path.join(tempfile.gettempdir(), self.DIRECTORY)
        if not os.path.exists(directory):
            os.makedirs(directory)
        file_path = os.path.join(directory, file_name)
        if create_file:
            if os.path.exists(file_path):
                os.remove(file_path)
            data = bytearray(size)
            with open(file_path, "w+b") as f:
                f.write(data)
        with open(file_path, "r+b") as f:
            # memory-map the file, size 0 means whole file
            self.accessor = mmap.mmap(f.fileno(), 0)
        self._file_path = file_path

    def get_int(self, offset: int) -> Tuple[int, int]:
        """
        Retrieves an integer from the shared memory
        :int offset: Where in the shared memory to retrieve the value
        :return: A tuple containing the value and the new offset
        """
        return struct.unpack_from("<i", self.accessor, offset)[0], offset + 4

    def get_float(self, offset: int) -> Tuple[float, int]:
        """
        Retrieves a float from the shared memory
        :int offset: Where in the shared memory to retrieve the value
        :return: A tuple containing the value and the new offset
        """
        return struct.unpack_from("<f", self.accessor, offset)[0], offset + 4

    def get_bool(self, offset: int) -> Tuple[bool, int]:
        """
        Retrieves a boolean from the shared memory
        :int offset: Where in the shared memory to retrieve the value
        :return: A tuple containing the value and the new offset
        """
        return struct.unpack_from("<?", self.accessor, offset)[0], offset + 1

    def get_string(self, offset: int) -> Tuple[str, int]:
        """
        Retrieves a string from the shared memory. The representation is an
        unsigned 8 bits value for the length followed by the bytes of the string
        in ASCII format.
        :int offset: Where in the shared memory to retrieve the value
        :return: A tuple containing the value and the new offset
        """
        string_len = struct.unpack_from("<B", self.accessor, offset)[0]
        byte_array = bytes(self.accessor[offset + 1 : offset + string_len + 1])
        result = byte_array.decode("ascii")
        return result, offset + string_len + 1

    def get_ndarray(
        self, offset: int, shape: Tuple[int, ...], t: np.dtype
    ) -> np.ndarray:
        """
        Retrieves a numpy array from the shared memory.
        :int offset: Where in the shared memory to retrieve the data
        :param shape: A tuple containing the shape of the expected array
        :param t: The dtype of the data to retrieve.
        :return: A new numpy array with the data from the shared memory
        """
        count = np.prod(shape)
        return np.frombuffer(
            buffer=self.accessor, dtype=t, count=count, offset=offset
        ).reshape(shape)

    def get_uuid(self, offset: int) -> Tuple[uuid.UUID, int]:
        """
        Retrieves a UUID from the shared memory
        :int offset: Where in the shared memory to retrieve the value
        :return: A tuple containing the value and the new offset
        """
        return uuid.UUID(bytes_le=self.accessor[offset : offset + 16]), offset + 16

    def set_int(self, offset: int, value: int) -> int:
        """
        Sets an integer into the shared memory.
        :int offset: The offset where to write the value
        :param value: The value to write
        :return: The new offset after writting the data
        """
        struct.pack_into("<i", self.accessor, offset, value)
        return offset + 4

    def set_float(self, offset: int, value: float) -> int:
        """
        Sets a float into the shared memory.
        :int offset: The offset where to write the value
        :param value: The value to write
        :return: The new offset after writting the data
        """
        struct.pack_into("<f", self.accessor, offset, value)
        return offset + 4

    def set_bool(self, offset: int, value: bool) -> int:
        """
        Sets a boolean into the shared memory.
        :int offset: The offset where to write the value
        :param value: The value to write
        :return: The new offset after writting the data
        """
        struct.pack_into("<?", self.accessor, offset, value)
        return offset + 1

    def set_string(self, offset: int, value: str) -> int:
        """
        Sets a string into the shared memory.
        :int offset: The offset where to write the value
        :param value: The value to write
        :return: The new offset after writting the data
        """
        string_len = len(value)
        struct.pack_into("<B", self.accessor, offset, string_len)
        offset += 1
        self.accessor[offset : offset + string_len] = value.encode("ascii")
        return offset + string_len

    def set_uuid(self, offset: int, value: uuid.UUID) -> int:
        """
        Sets a UUID into the shared memory.
        :int offset: The offset where to write the value
        :param value: The value to write
        :return: The new offset after writting the data
        """
        self.accessor[offset : offset + 16] = value.bytes_le
        return offset + 16

    def set_ndarray(self, offset: int, data: np.ndarray) -> None:
        """
        Sets a numpy array into the shared memory.
        :int offset: The offset where to write the value
        :param data: The value to write
        """
        bytes_data = data.tobytes()
        self.accessor[offset : offset + len(bytes_data)] = bytes_data

    def close(self) -> None:
        """
        Closes the shared memory reader and writter. This will not delete the file but
        remove access to it.
        """
        if self.accessor is not None:
            self.accessor.close()
            self.accessor = None  # type: ignore

    def delete(self) -> None:
        """
        Closes and deletes the shared memory file.
        """
        self.close()
        try:
            os.remove(self._file_path)
        except BaseException:
            pass

File: Runtime/Remote/BaseSharedMemory.cs
using System;
using System.IO;
using System.IO.MemoryMappedFiles;
using System.Text;
using Unity.Collections;
using Unity.Collections.LowLevel.Unsafe;

namespace Unity.AI.MLAgents
{
    unsafe internal class BaseSharedMemory
    {
        private const string k_Directory = "ml-agents";
        private MemoryMappedViewAccessor m_Accessor;
        private IntPtr m_AccessorPointer;
        private string m_FilePath;

        /// <summary>
        /// A bare bone shared memory wrapper that opens or create a new shared
        /// memory file and connects to it.
        /// </summary>
        /// <param name="fileName"> The name of the file. The file must be present
        /// in the Temporary folder (depends on OS) in the directory <see cref="k_Directory"/></param>
        /// <param name="createFile"> If true, the file will be created by the constructor.
        /// And error will be thrown if the file already exists.</param>
        /// <param name="size"> The size of the file to be created (will be ignored if
        /// <see cref="createFile"/> is true.</param>
        public BaseSharedMemory(string fileName, bool createFile, int size = 0)
        {
            var directoryPath = Path.Combine(Path.GetTempPath(), k_Directory);
            if (!Directory.Exists(directoryPath))
            {
                Directory.CreateDirectory(directoryPath);
            }
            m_FilePath = Path.Combine(directoryPath, fileName);
            if (createFile)
            {
                if (File.Exists(m_FilePath))
                {
                    throw new MLAgentsException($"The file {m_FilePath} already exists");
                }
                using (var fs = new FileStream(m_FilePath, FileMode.Create, FileAccess.Write))
                {
                    fs.Write(new byte[size], 0, size);
                }
            }
            long length = new System.IO.FileInfo(m_FilePath).Length;
            var mmf = MemoryMappedFile.CreateFromFile(
                File.Open(m_FilePath, FileMode.Open, FileAccess.ReadWrite, FileShare.ReadWrite),
                null,
                length,
                MemoryMappedFileAccess.ReadWrite,
                HandleInheritability.None,
                false
            );
            m_Accessor = mmf.CreateViewAccessor(0, length, MemoryMappedFileAccess.ReadWrite);
            mmf.Dispose();

            byte* ptr = (byte*)0;
            m_Accessor.SafeMemoryMappedViewHandle.AcquirePointer(ref ptr);
            m_AccessorPointer = new IntPtr(ptr);
        }

        /// <summary>
        /// Returns the integer present at the specified offset. The <see cref="offset"/> must be
        /// passed by reference and will be incremented to the next value.
        /// </summary>
        /// <param name="offset"> Where to read the value</param>
        /// <returns></returns>
        public int GetInt(ref int offset)
        {
            var result =  m_Accessor.ReadInt32(offset);
            offset += 4;
            return result;
        }

        /// <summary>
        /// Returns the float present at the specified offset. The <see cref="offset"/> must be
        /// passed by reference and will be incremented to the next value.
        /// </summary>
        /// <param name="offset"> Where to read the value</param>
        /// <returns></returns>
        public float GetFloat(ref int offset)
        {
            var result = m_Accessor.ReadSingle(offset);
            offset += 4;
            return result;
        }

        /// <summary>
        /// Returns the boolean present at the specified offset. The <see cref="offset"/> must be
        /// passed by reference and will be incremented to the next value.
        /// </summary>
        /// <param name="offset"> Where to read the value</param>
        /// <returns></returns>
        public bool GetBool(ref int offset)
        {
            var result = m_Accessor.ReadBoolean(offset);
            offset += 1;
            return result;
        }

        /// <summary>
        /// Returns the string present at the specified offset. The <see cref="offset"/> must be
        /// passed by reference and will be incremented to the next value.
        /// Note, the strings must be represented with : a sbyte (8-bit unsigned)
        /// indicating the size of the string in bytes followed by the bytes of the
        /// string (in ASCII format).
        /// </summary>
        /// <param name="offset"> Where to read the value</param>
        /// <returns></returns>
        public string GetString(ref int offset)
        {
            var length = m_Accessor.ReadSByte(offset);
            var bytes = new byte[length];
            offset += 1;
            m_Accessor.ReadArray(offset, bytes, 0, length);
            offset += length;
            return Encoding.ASCII.GetString(bytes);
        }

        /// <summary>
        /// Copies the values in the memory file at offset to the NativeArray <see cref="array"/>
        /// The <see cref="offset"/> must be passed by reference and will be incremented to the next value.
        /// <param name="offset"> Where to read the value</param>
        /// <param name="array"> The destination array</param>
        /// <param name="length"> The number of bytes that must be copied, must be less or equal
        /// to the capacity of the array.</param>
        /// <typeparam name="T"> The type of the NativeArray. Must be a struct.</typeparam>
        /// </summary>
        public void GetArray<T>(ref int offset, NativeArray<T> array, int length) where T : struct
        {
            if (length <= 0){
                return;
            }
            IntPtr src = IntPtr.Add(m_AccessorPointer, offset);
            IntPtr dst = new IntPtr(array.GetUnsafePtr());
            Buffer.MemoryCopy(src.ToPointer(), dst.ToPointer(), length, length);
            offset += length;
        }

        /// <summary>
        /// Returns the byte array present at the specified offset. The <see cref="offset"/> must be
        /// passed by reference and will be incremented to the next value.
        /// </summary>
        /// <param name="offset"> Where to read the value</param>
        /// <param name="length"> The size of the array in bytes</param>
        /// <returns></returns>
        public byte[] GetBytes(ref int offset, int length)
        {
            var result = new byte[length];
            m_Accessor.ReadArray(offset, result, 0, length);
            offset += length;
            return result;
        }

        /// <summary>
        /// Returns the integer present at the specified offset.
        /// </summary>
        /// <param name="offset"> Where to read the value.</param>
        /// <returns></returns>
        public int GetInt(int offset)
        {
            return m_Accessor.ReadInt32(offset);
        }

        /// <summary>
        /// Returns the float present at the specified offset.
        /// </summary>
        /// <param name="offset"> Where to read the value.</param>
        /// <returns></returns>
        public float GetFloat(int offset)
        {
            return m_Accessor.ReadSingle(offset);
        }

        /// <summary>
        /// Returns the boolean present at the specified offset.
        /// </summary>
        /// <param name="offset"> Where to read the value.</param>
        /// <returns></returns>
        public bool GetBool(int offset)
        {
            return m_Accessor.ReadBoolean(offset);
        }

        /// <summary>
        /// Returns the string present at the specified offset.
        /// Note, the strings must be represented with : a sbyte (8-bit unsigned)
        /// indicating the size of the string in bytes followed by the bytes of the
        /// string (in ASCII format).
        /// </summary>
        /// <param name="offset"> Where to read the value</param>
        /// <returns></returns>
        public string GetString(int offset)
        {
            var length = m_Accessor.ReadSByte(offset);
            var bytes = new byte[length];
            offset +=  1;
            m_Accessor.ReadArray(offset, bytes, 0, length);
            return Encoding.ASCII.GetString(bytes);
        }

        /// <summary>
        /// Copies the values in the memory file at offset to the NativeArray <see cref="array"/>
        /// <param name="offset"> Where to read the value</param>
        /// <param name="array"> The destination array</param>
        /// <param name="length"> The number of bytes that must be copied, must be less or equal
        /// to the capacity of the array.</param>
        /// <typeparam name="T"> The type of the NativeArray. Must be a struct.</typeparam>
        /// </summary>
        public void GetArray<T>(int offset, NativeArray<T> array, int length) where T : struct
        {
            IntPtr src = IntPtr.Add(m_AccessorPointer, offset);
            IntPtr dst = new IntPtr(array.GetUnsafePtr());
            Buffer.MemoryCopy(src.ToPointer(), dst.ToPointer(), length, length);
        }

        /// <summary>
        /// Returns the byte array present at the specified offset.
        /// </summary>
        /// <param name="offset"> Where to read the value</param>
        /// <param name="length"> The size of the array in bytes</param>
        /// <returns></returns>
        public byte[] GetBytes(int offset, int length)
        {
            var result = new byte[length];
            if (length > 0)
            {
                m_Accessor.ReadArray(offset, result, 0, length);
            }
            return result;
        }

        /// <summary>
        /// Sets the integer at the specified offset in the shared memory.
        /// </summary>
        /// <param name="offset"> The position at which to write the value</param>
        /// <param name="value"> The value to be written</param>
        /// <returns> The offset right after the written value.</returns>
        public int SetInt(int offset, int value)
        {
            m_Accessor.Write(offset, value);
            return offset + 4;
        }

        /// <summary>
        /// Sets the float at the specified offset in the shared memory.
        /// </summary>
        /// <param name="offset"> The position at which to write the value</param>
        /// <param name="value"> The value to be written</param>
        /// <returns> The offset right after the written value.</returns>
        public int SetFloat(int offset, float value)
        {
            m_Accessor.Write(offset, value);
            return offset + 4;
        }

        /// <summary>
        /// Sets the boolean at the specified offset in the shared memory.
        /// </summary>
        /// <param name="offset"> The position at which to write the value</param>
        /// <param name="value"> The value to be written</param>
        /// <returns> The offset right after the written value.</returns>
        public int SetBool(int offset, bool value)
        {
            m_Accessor.Write(offset, value);
            return offset + 1;
        }

        /// <summary>
        /// Sets the string at the specified offset in the shared memory.
        /// Note that the string is represented with a sbyte (8-bit unsigned) number
        /// encoding the length of the string and the string bytes in ASCII.
        /// </summary>
        /// <param name="offset"> The position at which to write the value</param>
        /// <param name="value"> The value to be written</param>
        /// <returns> The offset right after the written value.</returns>
        public int SetString(int offset, string value)
        {
            var bytes = Encoding.ASCII.GetBytes(value);
            int length = bytes.Length;
            m_Accessor.Write(offset, (sbyte)length);
            offset += 1;
            m_Accessor.WriteArray(offset, bytes, 0, length);
            offset += length;
            return offset;
        }

        /// <summary>
        /// Copies the values present in a NativeArray to the shared memory file.
        /// </summary>
        /// <param name="offset"> The position at which to write the value</param>
        /// <param name="array"> The NativeArray containing the data to write</param>
        /// <param name="length"> The size of the array in bytes</param>
        /// <typeparam name="T"> The type of the NativeArray. Must be a struct.</typeparam>
        /// <returns> The offset right after the written value.</returns>
        public int SetArray<T>(int offset, NativeArray<T> array, int length) where T : struct
        {
            IntPtr dst = IntPtr.Add(m_AccessorPointer, offset);
            IntPtr src = new IntPtr(array.GetUnsafePtr());
            Buffer.MemoryCopy(src.ToPointer(), dst.ToPointer(), length, length);
            return offset + length;
        }

        /// <summary>
        /// Sets the byte array at the specified offset in the shared memory.
        /// </summary>
        /// <param name="offset"> The position at which to write the value</param>
        /// <param name="value"> The value to be written</param>
        /// <returns> The offset right after the written value.</returns>
        public int SetBytes(int offset, byte[] value)
        {
            m_Accessor.WriteArray(offset, value, 0, value.Length);
            return offset + value.Length;
        }

        /// <summary>
        /// Closes the shared memory reader and writter. This will not delete the file but
        /// remove access to it.
        /// </summary>
        public void Close()
        {
            if (m_Accessor.CanWrite)
            {
                m_Accessor.SafeMemoryMappedViewHandle.ReleasePointer();
                m_Accessor.Dispose();
            }
        }

        /// <summary>
        /// Indicates wether the <see cref="BaseSharedMemory"/> has write access to the file.
        /// If it does not, it means the accessor was closed or the file was deleted.
        /// </summary>
        /// <value></value>
        protected bool CanEdit
        {
            get
            {
                return m_Accessor.CanWrite;
            }
        }

        /// <summary>
        /// Closes and deletes the shared memory file.
        /// </summary>
        public void Delete()
        {
            Close();
            File.Delete(m_FilePath);
        }
    }
}
